components/authz-service/server/v2/rules_test.go

```diff
@@ -157,6 +157,177 @@ func TestCreateRule(t *testing.T) {
+			assert.Nil(t, resp)
+		}},
+		{"if the rule id is empty, returns 'invalid argument'", func(t *testing.T) {
+			resp, err := cl.UpdateRule(ctx, &api.UpdateRuleReq{
+				Id:         "",
+				Name:       "empty id",
+				ProjectId:  "foo",
+				Conditions: apiConditions,
+				Type:       api.ProjectRuleTypes_NODE,
+			})
+			grpctest.AssertCode(t, codes.InvalidArgument, err)
+			assert.Nil(t, resp)
+		}},
+		{"if the rule id is invalid, returns 'invalid argument'", func(t *testing.T) {
+			resp, err := cl.UpdateRule(ctx, &api.UpdateRuleReq{
+				Id:         "no_underscores",
+				Name:       "any name",
+				ProjectId:  "foo",
+				Conditions: apiConditions,
+				Type:       api.ProjectRuleTypes_NODE,
+			})
+			grpctest.AssertCode(t, codes.InvalidArgument, err)
+			assert.Nil(t, resp)
+		}},
+		{"if there are no conditions, returns 'invalid argument'", func(t *testing.T) {
+			resp, err := cl.UpdateRule(ctx, &api.UpdateRuleReq{
+				Id:        "foo",
+				Name:      "foo rule",
+				ProjectId: "bar",
+				Type:      api.ProjectRuleTypes_NODE,
+			})
+			grpctest.AssertCode(t, codes.InvalidArgument, err)
+			assert.Nil(t, resp)
+		}},
+		{"if a rule with that id does not exist, returns 'not found'", func(t *testing.T) {
+			resp, err := cl.UpdateRule(ctx, &api.UpdateRuleReq{
+				Id:         "not-found",
+				Name:       "my other foo",
+				ProjectId:  "bar",
+				Conditions: apiConditions,
+				Type:       api.ProjectRuleTypes_NODE,
+			})
+			grpctest.AssertCode(t, codes.NotFound, err)
+			assert.Nil(t, resp)
+		}},
+		{"if the passed rule changes the project, returns 'failed precondition'", func(t *testing.T) {
+			id := "foo-rule"
+			addRuleToStore(t, store, id, "my foo rule", storage.Node, "foo-project", storageConditions)
+
+			resp, err := cl.UpdateRule(ctx, &api.UpdateRuleReq{
+				Id:         id,
+				Name:       "my other foo",
+				ProjectId:  "cannot-change",
+				Conditions: apiConditions,
+				Type:       api.ProjectRuleTypes_NODE,
+			})
+			grpctest.AssertCode(t, codes.FailedPrecondition, err)
+			assert.Nil(t, resp)
+		}},
+		{"if the updated rule passes conditions not compatible with the type, throw an error", func(t *testing.T) {
+			id := "foo-rule"
+			addRuleToStore(t, store, id, "my foo rule", storage.Event, "foo-project", storageConditions)
+
+			changedAPIConditions := []*api.Condition{
+				{
+					Type:     api.ProjectRuleConditionTypes_CHEF_TAGS,
+					Values:   []string{"should-break"},
+					Operator: api.ProjectRuleConditionOperators_EQUALS,
+				},
+			}
+
+			resp, err := cl.UpdateRule(ctx, &api.UpdateRuleReq{
+				Id:         id,
+				Name:       "my other foo",
+				ProjectId:  "foo-project",
+				Conditions: changedAPIConditions,
+				Type:       api.ProjectRuleTypes_EVENT,
+			})
+			grpctest.AssertCode(t, codes.InvalidArgument, err)
+			assert.Nil(t, resp)
+		}},
+		{"with valid rule data, returns no error and updates the rule in storage", func(t *testing.T) {
```

Comment:
This test is for a Node rule type with a condition **not** available for Event type.
A couple more useful tests:
Same test for **Node** rule type with a condition that **is** available for Event type.
Same test for **Event** rule type with a condition that is available for Node type.
(We cannot do the 4th possibility:  test for Event with condition not available for Node.)

(Also, if you have time, would be useful to have similar tests for CreateRule.)